Performance optimisation - Use `unordered_map` in `numbering` instead of `map`

thomasspriggs opened the pull request:
The commit which specifically gives the performance improvement is titled "Make `union_find` use hashes for numbering". The other commits are related refactoring.

Using `unordered_map` in `numbering` instead of `map`, changes `union_find`, which changes `arrays`, where the performance issue was seen. This change avoids slow lookups into `std::map` caused by long running `irept::operator<` calls. I have an example where cbmc spent approximately 70% of runtime in `irept::operator<` before this change and approximately 1% of runtime in hashing/ordering after this change.

Run time on example exercising arrays code before this PR -
real	2m39.018s
user	2m37.394s
sys	0m1.608s

Run time on example exercising arrays code after this PR -
real	0m30.897s
user	0m29.476s
sys	0m1.404s

Gprof before (top results up to `has_byte_operator` + `irept::hash()`) -
```
Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 44.90     21.08    21.08  2025113     0.00     0.00  irept::compare(irept const&) const
 26.84     33.68    12.60 4483321688     0.00     0.00  irept::number_of_non_comments(std::map<dstringt, irept, std::less<dstringt>, std::allocator<std::pair<dstringt const, irept> > > const&)
  6.54     36.75     3.07 52261322     0.00     0.00  Minisat::SimpSolver::addClause_(Minisat::vec<Minisat::Lit>&)
  3.43     38.36     1.61 52261322     0.00     0.00  Minisat::Solver::addClause_(Minisat::vec<Minisat::Lit>&)
  2.26     39.42     1.06    58881     0.00     0.00  has_byte_operator(exprt const&)
  0.06     46.03     0.03  3365106     0.00     0.00  irept::hash() const
```

Gprof after (top results up to `has_byte_operator` + `irept::hash()`) -
```
Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 27.40      3.11     3.11 52261322     0.00     0.00  Minisat::SimpSolver::addClause_(Minisat::vec<Minisat::Lit>&)
 13.74      4.67     1.56 52261322     0.00     0.00  Minisat::Solver::addClause_(Minisat::vec<Minisat::Lit>&)
  5.73      5.32     0.65 10600597     0.00     0.00  Minisat::SimpSolver::newVar(Minisat::lbool, bool)
  5.29      5.92     0.60    58881     0.00     0.00  has_byte_operator(exprt const&)
  0.18     10.51     0.02  3424407     0.00     0.00  irept::hash() const
```

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs reviewed (approved):
Good catch.  One thing to have a look at is the status of caching of the hashes in `dt`.  IIRC @tautschnig and @peterschrammel worked on this some time ago and I'm not sure if all of the options were enabled.

kroening commented:
It's odd that irep.h is a dependency of numbering.h.

I'd prefer an 'unordered_numbering' and an 'ordered_numbering', following the naming style of the STL. There may not be a use-case for the ordered numbering.

thomasspriggs committed:
Make `union_find` use hashes for numbering

This avoids slow lookups into `std::map` caused by long running
`irept::operator<` calls.

The type of hash to use is included as a template parameter in order to
make the data structure used to implement `union_find` more explicit.

Run time on example exercising arrays code before this commit -
real	2m39.018s
user	2m37.394s
sys	0m1.608s

Run time on example exercising arrays code after this commit -
real	0m30.897s
user	0m29.476s
sys	0m1.404s

hannes-steffenhagen-diffblue commented:
I agree with @kroening here, this can probably be done without special casing ireps. I believe @thomasspriggs is investigating if we can basically just use `hash_numbering` everywhere we can use `numbering` right now (I'd imagine so?).

thomasspriggs committed:
Reuse existing type definitions in `union_find`

Removing the duplicated types will enable changing the type of numbering
which backs union find, with fewer lines of code changed.

thomasspriggs committed:
Use `using` instead of `typedef` in `union_find`

Because this matches the current coding standard and because it improves
readability.

thomasspriggs committed:
Define new `numberingt`

This commit defines a new `numberingt`, which is equivalent to the
existing `hash_numbering` type. This new name and its template
parameters match the cbmc coding standards by ending the type names with
a letter 't'. It also supports using `std::hash` by default, which will
saves passing a hash function in the case where there is a standard hash
function defined for a given type of `keyt`.

The intention is to add this new type in this commit, swap over
the existing usages of numbering in the subsequent commits and then
finally remove the old variations of numbering. This will improve
performance in one instance and will improve maintainability in others.
Maintainability will be improved with a single version, because there
won't be multiple versions to maintain and it will take away the
possibility of accidentally choosing a variation which uses `irept` as
the key in the backing `std::map`.

thomasspriggs committed:
Use `numberingt` in place of `hash_numberingt`

Because the naming follows our coding standards and it reduces the
number of numbering variations to maintain.

thomasspriggs committed:
Use `numberingt` in place of `numbering`

Because the naming follows our coding standards and it reduces the
number of numbering variations to maintain. This does change the type of
backing container used in these places, however the performance should
be at least equivalent.

thomasspriggs committed:
Define `numberingt` as the implementation

There is no need for separate a separe type alias and implementaion if
numbering is always done using the same kind of map. Therefore we can
use a single name instead of two, which means less to maintain.

thomasspriggs commented:
I have now refactored this PR to avoid either `union_find` or `numbering` depending on `irep.h`. This PR removes the possibility of using an ordered map as the backing for numbering on the basis that this this variation is both unneeded and a potential performance foot-gun. This has the added benefit that we no longer need 3 different names for different numbering variations.

tautschnig commented:
@thomasspriggs What is the source of those `irept::operator<` calls? Is #1997 perhaps a way to actually address the root cause?

thomasspriggs commented:
@tautschnig For the particular example I have been examining, much of the overall runtime is spent in `src/solvers/flattening/arrays.cpp`. The `arrayst` class has a `union_find<exprt> arrays;`  field. This instance of `union_find<exprt>` was based on a `numbering<exprt>`, which in turn uses a `std::map<exprt, std::size_t>`, which in turn uses the `irept::operator<`.

From looking at the code I would say that the improvement in this PR looks unrelated to the improvement in #1997. As far as I can see, `arrays.cpp` is not using `sort_and_join`.

tautschnig commented:
@thomasspriggs Thank you for clarifying!
